utils/merge.go

```diff
@@ -0,0 +1,147 @@
+
+import (
+	"fmt"
+	"reflect"
+)
+
+// Merge will return a new struct/map/slice of the same type as base and patch, with patch merged into base.
+// Specifically, patch's values will be preferred except when patch's value is `nil`.
+// Restrictions/guarantees:
+//   - base and patch will not be modified
+//   - base and patch can be pointers or values
+//   - base and patch must be the same type
+//   - if slices are different, they will be merged according to the following rules:
+//       - start with elements from base (if any)
+//		 - then for each item, patch will overwrite if different
+//		 - and then add the extra items from patch (if any)
+//   - if maps are different, they will be merged according to the following rules:
+//       - keys in base are added first
+//       - keys in patch are added after, overwriting existing keys with values from patch
+//       - reference values (eg. slice/ptr/map/chan) will not be cloned
+//   - channel values are not supported at the moment
+//
+// Usage: callers need to cast the returned interface back into the original type, eg:
+// func mergeTestStruct(base, patch *testStruct) (*testStruct, error) {
+//     ret, err := merge(base, patch)
+//     if err != nil {
+//         return nil, err
+//     }
+//     retTS := ret.(testStruct)
+//     return &retTS, nil
+// }
+func Merge(base interface{}, patch interface{}) (interface{}, error) {
+	if reflect.TypeOf(base) != reflect.TypeOf(patch) {
+		return nil, fmt.Errorf("cannot merge different types. base type: %s, patch type: %s",
+			reflect.TypeOf(base), reflect.TypeOf(patch))
+	}
+
+	baseType := reflect.TypeOf(base)
+	baseVal := reflect.ValueOf(base)
+	patchVal := reflect.ValueOf(patch)
+	if baseType.Kind() == reflect.Ptr {
+		baseType = baseType.Elem()
+		baseVal = baseVal.Elem()
+		patchVal = patchVal.Elem()
+	}
+
+	ret := reflect.New(baseType)
+
+	val, ok := merge(baseVal, patchVal)
+	if ok {
+		ret.Elem().Set(val)
+	}
+	return ret.Elem().Interface(), nil
+}
+
+// merge recursively merges patch into base and returns the new struct, ptr, slice/map, or value
+func merge(base, patch reflect.Value) (reflect.Value, bool) {
+	baseType := base.Type()
+
+	switch baseType.Kind() {
+	case reflect.Struct:
+		merged := reflect.New(baseType).Elem()
+		for i := 0; i < base.NumField(); i++ {
+			if !merged.Field(i).CanSet() {
+				continue
+			}
+			val, ok := merge(base.Field(i), patch.Field(i))
+			if ok {
+				merged.Field(i).Set(val)
+			}
+		}
+		return merged, true
+
+	case reflect.Ptr:
+		mergedPtr := reflect.New(baseType.Elem())
+		if base.IsNil() && patch.IsNil() {
+			return mergedPtr, false
+		} else if base.IsNil() {
+			mergedPtr.Elem().Set(patch.Elem())
+			return mergedPtr, true
+		} else if patch.IsNil() {
+			mergedPtr.Elem().Set(base.Elem())
+			return mergedPtr, true
+		}
+		val, ok := merge(base.Elem(), patch.Elem())
+		if ok {
+			mergedPtr.Elem().Set(val)
+		}
+		return mergedPtr, true
+
+	case reflect.Slice:
+		// merge with these rules:
```

Comment:
Can you elaborate on when this might particular algorithm might be "useful"? I'm trying to think outside the box of just merging `*model.Config`, but I'm not sure I can imagine a case where this kind of "ovewrite" would be useful. In particular, if the slice was of type `[]*string`, we would happily replace non-`nil` values in the `base` slice with `nil` values from the `patch` slice, which runs counter to the overall idea of "merging" as such.

Reply:
Yes, that makes sense. I wasn't really sure how to handle slice and map merging here. I'll take it out and have an atomic overwrite if not nil.